lib/spack/spack/solver/asp.py

```diff
@@ -2016,33 +2032,60 @@ def _develop_specs_from_env(spec, env):
     spec.constrain(dev_info['spec'])
 
 
-#
-# These are handwritten parts for the Spack ASP model.
-#
-def solve(specs, dump=(), models=0, timers=False, stats=False, tests=False,
-          reuse=False):
-    """Solve for a stable model of specs.
-
-    Arguments:
-        specs (list): list of Specs to solve.
-        dump (tuple): what to dump
-        models (int): number of models to search (default: 0)
+class Solver(object):
+    """This is the main external interface class for solving.
+
+    It manages solver configuration and preferences in once place. It sets up the solve
+    and passes the setup method to the driver, as well.
+
+    Properties of interest:
+
+      ``reuse (bool)``
+        Whether to try to reuse existing installs/binaries
+
     """
-    driver = PyclingoDriver()
-    if "asp" in dump:
-        driver.out = sys.stdout
-
-    # Check upfront that the variants are admissible
-    for root in specs:
-        for s in root.traverse():
-            if s.virtual:
-                continue
-            spack.spec.Spec.ensure_valid_variants(s)
+    def __init__(self):
+        self.driver = PyclingoDriver()
+
+        # These properties are settable via spack configuration, and overridable
+        # by setting them directly as properties.
+        self.reuse = spack.config.get("concretizer:reuse", False)
 
-    setup = SpackSolverSetup()
-    return driver.solve(
-        setup, specs, dump, models, timers, stats, tests, reuse
-    )
+    def solve(self, specs, show=(), models=0, timers=False, stats=False, tests=False):
+        """
+        Arguments:
+          specs (list): List of ``Spec`` objects to solve for.
+          show (tuple): What information to print to the console while running.
+            Options are:
+            * asp: asp program text
+            * opt: optimization criteria for best model
+            * output: raw clingo output
+            * solutions: models found by asp program
+            * all: all of the above
```

Comment:
Is `all` still an option at this point? I think we parse `all` in the command and it's the lists of the others at this poinit

Reply:
actually none of these are used by `asp.py`. I'll consolidate it to just output.